# FREEDAThis is a tool Finding Rapidly Evolving Exons in De novo Assemblies.### Installation```pip install freeda```### Get startedHow to perform molecular evolution analysis using FREEDA:```Pythonfrom freeda import tblastnfrom freeda import exon_extractorfrom freeda import paml_launcherfrom freeda import paml_visualizerfrom freeda import structure_builderimport osdef freeda_pipeline(original_species=None, t=None, mafft_path=None):    # this needs to be a path to the "Data" folder    wdir = os.getcwd() + "/"    # reference species sequences: protein seq, cds, exons, gene (ex. Mus musculus)    if original_species == None:        original_species = "Mm"    # initial percent identity threshold for blast matches analysis    if t == None:        t = 30    # this needs to be a path where MAFFT aligner is installed    if mafft_path == None:        mafft_path = "/Users/damian/anaconda3/bin/mafft"    user_input1 = None    user_input2 = None    user_input3 = None    user_input4 = None    structure_overlay = True    structure_model_present = check_structure(wdir)    while user_input1 != "y" and user_input1 != "n":        user_input1 = input("(FREEDA) Should I run blast? (y / n)\n").lower()        if user_input1.lower() != "y" and user_input1.lower() != "n":            print("Please answer y or n\n")    while user_input2 != "y" and user_input2 != "n":        user_input2 = input("(FREEDA) Should I find exons? (y / n)\n").lower()        if user_input2.lower() != "y" and user_input2.lower() != "n":            print("Please answer y or n\n")    while user_input3 != "y" and user_input3 != "n":        user_input3 = input("(FREEDA) Should I run PAML? (y / n)\n").lower()        if user_input3.lower() != "y" and user_input3.lower() != "n":            print("Please answer y or n\n")    while user_input4 != "y" and user_input4 != "n":        user_input4 = input("(FREEDA) Should I run PyMOL? (y / n)\n").lower()        if user_input4.lower() != "y" and user_input4.lower() != "n":            print("Please answer y or n\n")    # make sure all structure prediction models are present    if user_input4 == "y" and not structure_model_present:        structure_overlay = False        print("...WARNING... (FREEDA) I cannot overlay adaptive sites until all proteins have structure predictions\n")        print("...WARNING... (FREEDA) I will run the pipeline skipping PyMOL\n")        print("...WARNING... (FREEDA) You can run PyMOL later on based on these results\n")    if user_input1 == "y":        blast_path = tblastn.run_blast(wdir, original_species)    else:        blast_path = wdir + "Blast_output/"    if user_input2 == "y":        result_path = exon_extractor.analyse_blast_results(wdir, blast_path,                                                           original_species, int(t), mafft_path)    if user_input3 == "y" and user_input2 == "n":        nr_of_tries = 1        while nr_of_tries <= 3:            user_input4 = input("----->  Indicate folder with results (no slashes or quotes): ")            result_path = wdir + user_input4 + "/"            if os.path.isdir(result_path) == False:                print("\n(FREEDA) I couldnt find your results folder (%s/3)" % nr_of_tries)                nr_of_tries += 1            else:                nr_of_tries = float("inf")                result_path = wdir + user_input4 + "/"                proteins, nr_of_species_total_dict, PAML_logfile_name, day = paml_launcher.analyse_final_cds(wdir,                                                                                                             original_species,                                                                                                             result_path,                                                                                                             mafft_path)                paml_visualizer.analyse_paml_results(wdir, result_path,                                                     proteins, nr_of_species_total_dict, original_species,                                                     PAML_logfile_name, day)    if user_input3 == "y" and user_input2 == "y":        proteins, nr_of_species_total_dict, PAML_logfile_name, day = paml_launcher.analyse_final_cds(wdir,                                                                                                     original_species,                                                                                                     result_path,                                                                                                     mafft_path)        paml_visualizer.analyse_paml_results(wdir, result_path,                                             proteins, nr_of_species_total_dict, original_species, PAML_logfile_name,                                             day)    if user_input1 == "n" and user_input2 == "n" and user_input3 == "n" and user_input4 == "y" and structure_model_present:        nr_of_tries = 1        while nr_of_tries <= 3:            user_input4 = input("----->  Indicate folder with results (no slashes or quotes): ")            result_path = wdir + user_input4 + "/"            if os.path.isdir(result_path) == False:                print("\n(FREEDA) I couldnt find your results folder (%s/3)" % nr_of_tries)                nr_of_tries += 1            else:                nr_of_tries = float("inf")                result_path = wdir + user_input4 + "/"                structure_builder.run_pymol(wdir, result_path)                # indicate that structure overlay was already run                structure_overlay = False    if structure_model_present and structure_overlay:        structure_builder.run_pymol(wdir, result_path)    print("\nYou reached the end of FREEDA pipeline.")def check_structure(wdir):    with open("proteins.txt", "r") as f:        file = f.readlines()        for line in file:            protein = line.rstrip("\n")            structures_path = wdir + "Structures"            model_path = structures_path + "/" + protein + "/model1.pdb"            if os.path.isfile(model_path):                structure_model_present = True            else:                structure_model_present = False                print("...WARNING... Structure prediction for protein: %s DOES NOT EXIST" % protein)    # do not return any script if even one model is not present    return structure_model_presentif __name__ == '__freeda_pipeline__':    import argparse    parser = argparse.ArgumentParser(description="Run FREEDA pipeline")    parser.add_argument("-os", "--original_species",                        help="specify reference organism (ex. Mm for Mus musculus)", type=str,                        required=False)    parser.add_argument("-t", "--blast_threshold",                        help="specify percentage identity threshold for blast (ex. 30; default)", type=int,                        required=True)    parser.add_argument("-mafft", "--mafft_path",                        help="specify path to mafft aligner (ex. /Users/user/anaconda/bin/mafft)", type=str,                        required=True)    args = parser.parse_args()    freeda_pipeline(original_species=args.original_species, t=args.blast_threshold, mafft_path=args.mafft_path)```