
<!DOCTYPE html>

<html lang="python">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Troubleshooting &#8212; FREEDA 0.1 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Development" href="development.html" />
    <link rel="prev" title="Genomic assemblies used" href="assemblies.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
    <link rel="apple-touch-icon" href="_static/./freeda_logo.png" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/./freeda_logo.png" alt="Logo"/>
    
  </a>
</p>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="getstarted.html">How to get started</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="assemblies.html">Genomic assemblies used</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Troubleshooting</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#fatal-errors-action-needed">Fatal errors (action needed)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#warnings-no-action-needed">Warnings (no action needed)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#problems-with-virtualbox">Problems with VirtualBox</a></li>
<li class="toctree-l2"><a class="reference internal" href="#other-problems">Other problems</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Development</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="assemblies.html" title="previous chapter">Genomic assemblies used</a></li>
      <li>Next: <a href="development.html" title="next chapter">Development</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="troubleshooting">
<h1>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Permalink to this heading">¶</a></h1>
<p>The most common error comes from unstable Internet connection, especially during downloading genomes (first run). Simply close and run FREEDA again.</p>
<dl class="simple">
<dt>In case your issue is not covered here please send (<strong>damiandudka0&#64;gmail.com</strong>):</dt><dd><ul class="simple">
<li><p>print-screen of the GUI</p></li>
<li><p><em>FREEDA-current-date.log</em> file (in &quot;Raw_data&quot; folder or in the folder indicated by &quot;Set directory&quot; see: <a class="reference internal" href="usage.html#understanding-results"><span class="std std-ref">Usage</span></a>)</p></li>
<li><p>(if present) <em>PAML-current-date.log</em> file (in &quot;Raw_data&quot; folder or in the folder indicated by &quot;Set directory&quot; see: <a class="reference internal" href="usage.html#understanding-results"><span class="std std-ref">Usage</span></a>)</p></li>
</ul>
</dd>
</dl>
<div class="section" id="fatal-errors-action-needed">
<h2>Fatal errors (action needed)<a class="headerlink" href="#fatal-errors-action-needed" title="Permalink to this heading">¶</a></h2>
<p><strong>...FATAL ERROR... : Something went wrong (see below). Send screen shot to : Damian Dudka -&gt; damiandudka0&#64;gmail.com</strong></p>
<blockquote>
<div><p><em>REASON:</em> Follows most unknown errors causing FREEDA to crash.</p>
<p><em>SOLUTION:</em> Please attach a screen shot and &quot;FREEDA-current-date.log&quot; and (if present) &quot;PAML-current-date.log&quot;. They will be either in &quot;Raw_data&quot; folder or in the main folder indicated within the GUI (&quot;Set directory&quot;).</p>
</div></blockquote>
<p><strong>...FATAL ERROR... : PyMOL not found in the PATH, the Applications folder, or the working directory.</strong></p>
<blockquote>
<div><p><em>REASON:</em> This is most likely a iOS user error installing PyMOL outside of the Applications folder.</p>
<p><em>SOLUTION:</em> Move the PyMOL .app to Applications.</p>
</div></blockquote>
<p><strong>...FATAL ERROR... : Partial or empty genome file : GENOME.fasta -&gt; likely Internet connection was disrupted -&gt; please run FREEDA again...</strong></p>
<blockquote>
<div><p><em>REASON:</em> The most common error at first run - almost always caused by unstable Internet connection or closing the app when downloading the genomes at first run.</p>
<p><em>SOLUTION:</em> Simply re-run the app.</p>
</div></blockquote>
<p><strong>...FATAL ERROR... : Genome : GENOME blast databases failed to build (likely interrupted) -&gt; rerun the pipeline...</strong></p>
<blockquote>
<div><p><em>REASON:</em> Most likely the app crashed or was closed while making local BLAST databases.</p>
<p><em>SOLUTION:</em> Try re-running the app.</p>
</div></blockquote>
<p><strong>...FATAL ERROR... : Genome : GENOME blast databases were built partially (likely interrupted) -&gt; rerun the pipeline...</strong></p>
<blockquote>
<div><p><em>REASON:</em> Similar to the previous error. FREEDA is expecting a certain size of the generated databases. Partial files may be present when the run was interrupted.</p>
<p><em>SOLUTION:</em> Try re-running the app.</p>
</div></blockquote>
<p><strong>...FATAL ERROR... : Repetitive coding sequence detected in GENE (min 80bp repeat) -&gt; cannot reliably analyze this GENE</strong></p>
<blockquote>
<div><p><em>REASON:</em> FREEDA's ability to find orthologous exons relies on accurate alignment to the reference coding sequence. Repetitive sequences are notoriously difficult to align, therefore FREEDA will not try to analyze genes with more than 80bp-long repeats in coding sequence (e.g. Tacc3 in rodents).</p>
<p><em>SOLUTION:</em> Unfortunately this gene cannot be analyzed reliably. If you still want to try I can run the analysis on my end and let you know if its reliable: <a class="reference external" href="mailto:damiandudka0&#37;&#52;&#48;gmail&#46;com">damiandudka0<span>&#64;</span>gmail<span>&#46;</span>com</a></p>
</div></blockquote>
<p><strong>...FATAL ERROR... : GENE is transposable element instead of a protein coding gene -&gt; check gene name for SPECIES here: ensembl.org -&gt; correct gene name and click Analyze again...</strong></p>
<blockquote>
<div><p><em>REASON:</em> Provided gene name does not encode a protein (but its a trasposable element in this case).</p>
<p><em>SOLUTION:</em> Check out the link FREEDA gives you and ensure you know the name of the protein coding gene you are interested in. ALWAYS use the release FREEDA gives you the link to.</p>
</div></blockquote>
<p><strong>...FATAL ERROR... : No reliable coding sequence annotation detected for GENE</strong></p>
<blockquote>
<div><p><em>REASON:</em> This is usually caused by unreliable annotation of the gene of interest - mostly concerning dog and chicken assemblies.</p>
<p><em>SOLUTION:</em> Verify your gene annotation that FREEDA used under the link provided</p>
</div></blockquote>
<p><strong>...FATAL ERROR... : Gene names [GENE] do not exist in reference assembly -&gt; check gene name for SPECIES here: ensembl.org -&gt; correct gene name and click Analyze again...</strong></p>
<blockquote>
<div><p><em>REASON:</em> This is usually caused by unreliable annotation of the gene of interest - mostly concerning dog and chicken assemblies.</p>
<p><em>SOLUTION:</em> Make sure there is a known coding sequence for this gene (e.g. visit Ensembl database - <a class="reference external" href="https://useast.ensembl.org/index.html">https://useast.ensembl.org/index.html</a></p>
</div></blockquote>
<p><strong>...FATAL ERROR... : Input data generation FAILED for GENE - please remove GENE from analysis -&gt; exiting the pipeline now...</strong></p>
<blockquote>
<div><p><em>REASON:</em> Unreliable input generation - usually following another FATAL_ERROR (e.g. no coding transcript found)</p>
<p><em>SOLUTION:</em> Make this is not a pseudogene (e.g. visit Ensembl database - <a class="reference external" href="https://useast.ensembl.org/index.html">https://useast.ensembl.org/index.html</a></p>
</div></blockquote>
<p><strong>...FATAL ERROR... : &gt;= 5 species show no matches above blast threshold: 60% Gene: GENE -&gt; please run FREEDA again with Exclude species option (copy paste this: Pt Gg)</strong></p>
<blockquote>
<div><p><em>REASON:</em> FREEDA uses a default 60% (vertebrates) or 30% (flies) treshold for identity when blasting coding sequences. In case of very divergent genes that threshold might not be met. If you used &quot;Common domains expected&quot; advanced option that increases the threshold to 80% (vertebrates) or 60% (flies) increasing the risk of no matches found for a given species. It might also be due to a gene loss in some species.</p>
<p><em>SOLUTION:</em> Try the &quot;Exclude species&quot; option using two-letter (or three-letter for flies) code for each genome (e.g., type Pt Gg for <em>Pan troglodytes</em> and <em>Gorilla gorilla</em>; or Dsi Der for <em>Drosophila simulans</em> and <em>Drosophila erecta</em>). You can also try to indicate a subgroup (hominoidea, catarrhini, caniformia, melanogaster) to avoid blasting against more divergent species.</p>
</div></blockquote>
<p><strong>...FATAL ERROR... : Too few genomes were selected (&lt; 6)</strong></p>
<blockquote>
<div><p><em>REASON:</em> Similar to the previous error. FREEDA expects coding sequences from at least 6 species (genomic assemblies) to be able to run the analysis. This is most likely to happen if you select melanogaster as subgroup (maximum 9 species available).</p>
<p><em>SOLUTION:</em> Make sure not to select &quot;Common domains expected&quot; and do not select melanogaster subgroup (allowing search in more divergent flies too).</p>
</div></blockquote>
</div>
<div class="section" id="warnings-no-action-needed">
<h2>Warnings (no action needed)<a class="headerlink" href="#warnings-no-action-needed" title="Permalink to this heading">¶</a></h2>
<p><strong>AlphaFold structure not found or not matching Ensembl input collected</strong></p>
<blockquote>
<div><p><em>REASON:</em> FREEDA will still run the analysis but without mapping residues onto structure</p>
<img alt="_images/GUI_events_No_structure.png" src="_images/GUI_events_No_structure.png" />
</div></blockquote>
<p><strong>Questionable alignment of a single exon</strong></p>
<blockquote>
<div><p><em>REASON:</em> FREEDA performs additional checks (blue) for a divergent exon and accepts or rejects it</p>
<img alt="_images/GUI_events_single_exon_warning.png" src="_images/GUI_events_single_exon_warning.png" />
</div></blockquote>
<p><strong>Coding sequence is not in frame</strong></p>
<blockquote>
<div><p><em>REASON:</em> Either some exons are missing (not the case in example below) or single indels
are present (e.g. sequencing errors). FREEDA may either remove this sequence from analysis
or remove the indels to force conserved alignment.</p>
<img alt="_images/GUI_events_CDS_not_in_frame.png" src="_images/GUI_events_CDS_not_in_frame.png" />
</div></blockquote>
<p><strong>Early STOP codon detected in final coding sequence</strong></p>
<blockquote>
<div><p><em>REASON:</em> Coding sequences in the final alignment should not have any STOP codons anymore,
except those introduced by imperfect alignment. This is rare and usually occurs in highly
divergent genes. FREEDA removes the STOP codon forcing a conserved alignment.</p>
<img alt="_images/GUI_STOP_codon_warning.png" src="_images/GUI_STOP_codon_warning.png" />
</div></blockquote>
<p><strong>Failed check comparing cloned sequence to annotated one for most distant species</strong></p>
<blockquote>
<div><p><em>REASON:</em> This is a sanity check. Usually &lt;95% identity suggests that alternative exons
are used (check supported only for rodents and carnivores).</p>
<img alt="_images/GUI_events_no_matching.png" src="_images/GUI_events_no_matching.png" />
</div></blockquote>
</div>
<div class="section" id="problems-with-virtualbox">
<h2>Problems with VirtualBox<a class="headerlink" href="#problems-with-virtualbox" title="Permalink to this heading">¶</a></h2>
<div class="section" id="bit-operating-systems-won-t-show-up">
<h3>64-bit Operating Systems Won't Show Up<a class="headerlink" href="#bit-operating-systems-won-t-show-up" title="Permalink to this heading">¶</a></h3>
<p>Your Windows system may already be using another Virtual Machine system called HyperV.</p>
</div>
</div>
<div class="section" id="other-problems">
<h2>Other problems<a class="headerlink" href="#other-problems" title="Permalink to this heading">¶</a></h2>
<div class="section" id="pymol-desktop-file-may-not-work">
<h3>PyMOL desktop file may not work<a class="headerlink" href="#pymol-desktop-file-may-not-work" title="Permalink to this heading">¶</a></h3>
<p>Ubuntu 22.04 LTS version does not seem to support giving launching permissions to Desktop files
but that might change with future updates. For now just double-click on .pse files to open structures
in PyMOL on Linux.</p>
</div>
<div class="section" id="pymol-tar-unpacking-cannot-create-symlink">
<h3>PyMOL Tar Unpacking &quot;Cannot Create Symlink&quot;<a class="headerlink" href="#pymol-tar-unpacking-cannot-create-symlink" title="Permalink to this heading">¶</a></h3>
<p>Your data folder may be in a system partition format that does not allow symbolic links. As a work around, you can install PyMOL yourself from a package manager, software application, or the open source PyMOL repository: <cite>https://github.com/schrodinger/pymol-open-source &lt;https://github.com/schrodinger/pymol-open-source&gt;</cite>.</p>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2022, Damian Dudka, R. Brian Akins.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 5.3.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/troubleshooting.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>